marker_utils/src/visitor.rs

```diff
@@ -293,3 +367,72 @@ pub fn traverse_expr<'ast, B>(
 
     ControlFlow::Continue(())
 }
+
+/// This trait is implemented for nodes, that can be traversed by a [`Visitor`].
+pub trait Traverseable<'ast, B> {
+    /// This calls the `traverse_*` function for the implementing node.
+    fn traverse(self, cx: &'ast AstContext<'ast>, visitor: &mut dyn Visitor<B>) -> ControlFlow<B>;
+}
+
+/// This macro implements the [`Traverseable`] trait for a given node that implements `Copy`
+macro_rules! impl_traversable_for {
+    ($ty:ty, $func:ident) => {
+        impl<'ast, B> Traverseable<'ast, B> for $ty {
+            fn traverse(self, cx: &'ast AstContext<'ast>, visitor: &mut dyn Visitor<B>) -> ControlFlow<B> {
+                $func(cx, visitor, self)
+            }
+        }
+    };
+}
+
+impl_traversable_for!(ExprKind<'ast>, traverse_expr);
+impl_traversable_for!(StmtKind<'ast>, traverse_stmt);
+impl_traversable_for!(ItemKind<'ast>, traverse_item);
+impl_traversable_for!(&'ast Body<'ast>, traverse_body);
+
+/// This function calls the given closure for every expression in the node. This
+/// traversal will not enter any nested bodies.
+///
+/// This function is a simple wrapper around the [`Visitor`] trait, that is good
+/// for most use cases. For example, the following code counts the number of `if`s
+/// in a body:
+///
+/// ```
+/// # use marker_api::prelude::*;
+/// # use std::ops::ControlFlow;
+/// # use marker_utils::visitor::for_each_expr;
+/// fn count_ifs<'ast>(cx: &'ast AstContext<'ast>, body: &'ast Body<'ast>) -> u32 {
+///     let mut count = 0;
+///     let _: Option<()> = for_each_expr(
+///         cx,
+///         body,
+///         |expr| {
+///             if matches!(expr, ExprKind::If(_)) {
+///                 count += 1;
+///             }
+///             ControlFlow::Continue(())
```

Comment:
Hm, I wish it could be an iterator 🤔

Reply:
I don't see a simple way to turn this into an iterator without either first collecting every node into a vector or using generators. Out of interested, is there a specific use case, you thought of?

Reply:
Not a specific one. Iterators are undoubtedly harder to implement but are easier to use because you can cain a whole pipeline of combinators for them. While the callback style for-each with a `ControlFlow` return is quite procedural-looking, and barebones.

For example, RA's `rowan` crate [provides iterators](https://docs.rs/rowan/latest/rowan/api/struct.SyntaxNode.html#method.children) to traverse the syntax node tree, and it's been a pleasure to use that kind of API.